[doc] Add policy for merging and promotion (#3187)
habitat-sh/core-plans: docs/dev/policy_documents/merging-and-promoting-conventions.md

# Merging and Promoting Conventions

## Summary
[summary]: #summary

This policy defines a 'fast-track' merge and promote policy for a subset of core-plans.  

## Motivation
[motivation]: #motivation

> If someone who contributes heavily to core plans, has a pull request for a minor change open for 7days +, with all tests passing, it would be good if the same person can also merge and promote if there is a lack of peer reviews. One contributor (Graham Wilson, predominant) has hundreds of updates in his backlog, but has been holding back because of delays in the review process.

Reviewer: ```suggestion
Our code review process is currently slow, frustrating core-plan maintainers who are donating their personal time or their employer's time to the project. One contributor (Graham Wilson, predominant) has hundreds of updates in his backlog, but has been holding back because of delays in the review process.

If someone who contributes heavily to core plans has a pull request (PR) for a minor change open for 7+ days, the automated testing already in place and the small scope of the change makes the limited value of additional code-review small compared to the cost it is imposing on some maintainers.  In this situation, allowing veteran maintainers to use their own judgement and immediately merge the PR will help ensure core-plan maintainers can make the most of the time they spend on this project.
```

Author: Perfect!

## Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> What plan is eligable for the fast-track criteria:

Reviewer: ```suggestion
What pull requests are eligible for the fast-track criteria:
```

Author: Thanks

> * Only identified 'veteran' core-plan contributor's have permission to use this fast-track policy.  This list will be agreed and specified in the MAINTAINERS.md

Reviewer: ```suggestion
* Only identified 'veteran' core-plan contributors have permission to use this fast-track policy.  This list will be agreed and specified in the MAINTAINERS.md
```

> * This fasttrack will only apply to non-baseplan PRs and PRs that do not have a label 'DO-NOT-MERGE' or are otherwise marked for caution.

Reviewer: ```suggestion
* This fast-track will only apply to non-base plan PRs and PRs that do not have a label 'DO-NOT-MERGE' or are otherwise marked for caution.
```

> * If no feedback has been added to the PR after 7+ days, then the veteran contributor can merge and promote the PR.

Reviewer: Suggested Addition:

* The PR consists of minor changes. For example, version bumps, style improvements, bug fixes that don't change the scope of the plan.

Minor style suggestion: 

I think it might be better to rephrase these eligibility requirements in the form of positive assertions that can be checked:

- PR was opened by 'veteran' core-plan contributor. Veteran core-plan contributors are specified in MAINTAINERS.md;
- PR is for a non-base plan and does not have the 'DO-NOT-MERGE' label;
- All automated testing has passed; and,
- PR has not received feedback for 7+ days.

Author: Thanks @stevendanna. Will add these.

## Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> Delete this section?

Reviewer: I agree we can merge this and the previous section for this one.

Author: Deleted.

## Drawbacks
[drawbacks]: #drawbacks

> What are the drawbacks if any?

Reviewer: Drawback include (1) less awareness amongst maintainers about what changes have occurred recently, (2) opportunity for bugs to enter that would have been caught in code review, (3) decreases pressure to improve the code-review process that has led to long delays.

Author: Updating according to these comments

# Merging and Promoting Conventions

## Summary
[summary]: #summary

This policy defines a 'fast-track' merge and promote policy for a subset of core-plans.  

## Motivation
[motivation]: #motivation

Our code review process is currently slow, frustrating core-plan maintainers who are donating their personal time or their employer's time to the project. One contributor (Graham Wilson, predominant) has hundreds of updates in his backlog, but has been holding back because of delays in the review process.

> If someone who contributes heavily to core plans has a pull request (PR) for a minor change open for 7+ days, the automated testing already in place and the small scope of the change makes the limited value of additional code-review small compared to the cost it is imposing on some maintainers.  In this situation, allowing 'aprover' maintainers to use their own judgement and immediately merge the PR will help ensure core-plan maintainers can make the most of the time they spend on this project.

Reviewer: ```suggestion
If someone who contributes heavily to core plans has a pull request (PR) for a minor change open for 7+ days, the automated testing already in place and the small scope of the change makes the value of additional code-review small compared to the cost it is imposing on some maintainers.  In this situation, allowing 'approver' maintainers to use their own judgement and immediately merge the PR will help ensure core-plan maintainers can make the most of the time they spend on this project.
```
